Migration of AP_HAL_ESP32 targets from esp-idf 4.4 to 5.3, take two

arg7 opened the pull request:
Main focus of this PR:

- Migration to the current ESP-IDF build environment (v5.3)

Additional updates:

- Fixed gcc 13 compilation issues
- Fixed WDT in Scheduler
- Updated README.md

See commit messages for details.

Tested building of ChibiOS, ESP32 and ESP32S3 targets.

Tested ESP32S3 build on ESP32S3DevKit board.

Best regards,
Ing. A. Romanov

tpwrules reviewed (dismissed):
This is looking a lot better, thanks for the effort! I have added comments to things I am qualified to comment on.

There are still issues with the commit structure however:
* Commits must be split up so that each commit only touches one subsystem (`Tools`, `AP_Filesystem`, `SITL`, `AP_HAL_ESP32` etc) and the commit message must be prefixed with the subsystem name (e.g. `AP_Filesystem: fix build with ESP32 and GCC13`)
* Please avoid mixing changes up in commits, for example there is one commit that both updates the printfs and changes the watchdog behavior. This needs to be two separate commits
* You can put more detailed messages and reasoning after the basic description. (e.g. `AP_Filesystem: fix build with ESP32 and GCC13 <next paragraph> ensure subclass definitions match superclass`)
* Speaking of the printfs, I think the `DBG_PRINTF` changes need to be their own PR. There are some I'm not sure about and the changes are distracting

arg7 commented on Tools/scripts/esp32_get_idf.sh:
```diff
@@ -23,40 +28,48 @@ else
-        # check if we've got v4.4 checked out, only this version of esp_idf is tested and works?
-        
+        git clone -b 'release/v5.3'  https://github.com/espressif/esp-idf.git esp_idf
+        git checkout $CHECKOUT
```
Such commits IDs were used before and not being git regular user I didn't grasp the point.
I'll fix as suggested.

arg7 commented on Tools/scripts/esp32_get_idf.sh:
```diff
@@ -23,40 +28,48 @@ else
+cd modules/esp_idf
 git submodule update --init --recursive
+
+echo
+echo "apply patch for SOC_RMT_SUPPORT_RX_PINGPONG bug"
+sed -i 's/^#define SOC_RMT_SUPPORT_RX_PINGPONG.*/#define SOC_RMT_SUPPORT_RX_PINGPONG 0 \/* esps32s3 does not support PING-PONG feature *\//' ./components/soc/esp32s3/include/soc/soc_caps.h
+
```
I thought that fixing this bug at the source is a good idea.

In nex PR regading RMT fix I'll add backup:

#ifdef CONFIG_IDF_TARGET_ESP32S3
  #if SOC_RMT_SUPPORT_RX_PINGPONG == 1
    #undef SOC_RMT_SUPPORT_RX_PINGPONG
    #define SOC_RMT_SUPPORT_RX_PINGPONG 0
  #endif
#endif

davidbuzz replied:
I think your suggested change here is better.

arg7 replied:
ok, I'll remove this patch from PR

arg7 commented on Tools/environment_install/install-prereqs-arch.sh:
```diff
@@ -66,6 +66,8 @@ fi
 if [[ $DO_PYTHON_VENV_ENV -eq 1 ]]; then
     echo "$SOURCE_LINE" >> ~/.bashrc
+else
+    echo "Use $SOURCE_LINE to ativate Ardupilot venv"
```
ok

arg7 commented on Tools/environment_install/install-prereqs-openSUSE-Tumbleweed.sh:
```diff
@@ -102,6 +102,8 @@ if ! grep -Fxq "$SOURCE_LINE" ~/.bashrc; then
     if [[ $DO_PYTHON_VENV_ENV -eq 1 ]]; then
         echo $SOURCE_LINE >> ~/.bashrc
+    else
+        echo "Use $SOURCE_LINE to ativate Ardupilot venv"
```
ok

arg7 commented on Tools/environment_install/install-prereqs-ubuntu.sh:
```diff
@@ -399,6 +393,8 @@ if [ -n "$PYTHON_VENV_PACKAGE" ]; then
     if [[ $DO_PYTHON_VENV_ENV -eq 1 ]]; then
         echo $SOURCE_LINE >> ~/$SHELL_LOGIN
+    else
+	echo "Use $SOURCE_LINE to ativate Ardupilot venv"
```
ok

arg7 commented on libraries/AP_HAL_ESP32/HAL_ESP32_Class.cpp:
```diff
@@ -26,19 +26,22 @@
 #include "RCInput.h"
 #include "RCOutput.h"
 #include "Storage.h"
+
 #include "AnalogIn.h"
+
 #include "Util.h"
```
ok, I will remove white space.

arg7 commented on libraries/AP_Filesystem/posix_compat.h:
```diff
@@ -58,7 +58,10 @@ long apfs_ftell(APFS_FILE *stream);
 APFS_FILE *apfs_freopen(const char *pathname, const char *mode, APFS_FILE *stream);
 int apfs_remove(const char *pathname);
 int apfs_rename(const char *oldpath, const char *newpath);
+
+#ifndef ESP_PLATFORM
 char *tmpnam(char s[L_tmpnam]);
+#endif
```
Yes, idf 5.3 redefines this function.

arg7 commented on libraries/AP_HAL_ESP32/Scheduler.cpp:
```diff
@@ -243,15 +229,15 @@ void IRAM_ATTR Scheduler::delay_microseconds(uint16_t us)
     }
     if (_num_timer_procs >= ESP32_SCHEDULER_MAX_TIMER_PROCS) {
-        printf("Out of timer processes\n");
+        DBG_PRINTF("Out of timer processes\n");
```
if it is error, then it will be better to rewrite it as GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "Out of timer processes, limit was: %d\n", ESP32_SCHEDULER_MAX_TIMER_PROCS);

Otherwise nobody will see it.

tpwrules replied:
You’re not wrong. Like I said in the main comment I think the printf changes should be moved to a separate PR for more focused discussion

arg7 replied:
ok, i reverted DBG_PRINT macro, I'll include it in the next PR.

arg7 commented on Tools/environment_install/install-prereqs-ubuntu.sh:
```diff
@@ -375,16 +375,10 @@ fi
-    PYTHON_VENV_PACKAGE=python3.12-venv
-fi
+PYTHON_VENV_PACKAGE="python$(python3 --version | awk '{ print $2 }' | cut -d '.' -f 1,2)-venv"
+VENV_LOOKUP=$(apt search $PYTHON_VENV_PACKAGE 2>/dev/null| grep venv)
 
-if [ -n "$PYTHON_VENV_PACKAGE" ]; then
+if [[ $VENV_LOOKUP ]]; then
```
There was a bug, which skipped installation of VENV on quite popular Ubuntu 22.04 code name jammy, which I fixed.
Since it was related to installation experience, I included it in this PR.

arg7 commented on libraries/SITL/SIM_EFI_MegaSquirt.cpp:
```diff
@@ -107,7 +107,7 @@ void EFI_MegaSquirt::update()
             send_table();
         } else {
-            printf("BAD EFI CRC: 0x%08x 0x%08x\n", crc, crc2);
+            printf("BAD EFI CRC: 0x%08x 0x%08x\n", (unsigned int)crc, (unsigned int)crc2);
```
But it prevents compile with idf 5.3, because it uses gcc 13.
I've discussed inclusion of this change with Buzz, he was agree.

tpwrules replied:
Yes, the change is correct and it’s OK to be in this PR, it just needs to be in its own commit.

davidbuzz replied:
this (unsigned int) fix has now been brought into master as its own bugfix commit

arg7 replied:
Ok, I'll remove it from PR.

arg7 commented on Tools/environment_install/install-prereqs-ubuntu.sh:
```diff
@@ -1,7 +1,7 @@
 echo "---------- $0 start ----------"
 set -e
-set -x
+#set -x
```
Only on Ubuntu this install-prereq script uses debug trace for no apparent reason.
Also this script is overly complicated, can improved a lot.

khancyr replied:
this is usefull on CI to see where it failed

arg7 replied:
I've seen that other install-prereqs-* scripts doesn't emit trace info.
So I thought that once debugged, it is not useful anymore.

arg7 commented:
This PR contains only few necessary changes to files outside of ESP32 target which are necessary to compile Ardupilot on new gcc 13 included in idf 5.3,

I've already discussed their inclusion with Buzz and Thomas, it was ok with them.

tpwrules commented:
The contents look generally good but the commit messages still need fixing. I can do it if you would like?

I can also submit the fixes in other modules as a separate PR to (very slightly) shrink this one if you would prefer.

arg7 commented:
Hi Thomas,
sure you can do as you see fit.

davidbuzz commented on libraries/AP_Filesystem/AP_Filesystem_ESP32.cpp:
```diff
@@ -42,15 +42,15 @@ int AP_Filesystem_ESP32::close(int fd)
 }
 
-ssize_t AP_Filesystem_ESP32::write(int fd, const void *buf, size_t count)
+int32_t AP_Filesystem_ESP32::write(int fd, const void *buf, uint32_t count)
```
this has now been brought into master as its own bugfix commit

arg7 replied:
Ok, I'll remove it from PR.

davidbuzz replied:
re-basing your PR should achieve this

davidbuzz reviewed (commented):
I'm generally in-support of 99% of the esp32 HAL changes.
it needs to be be squashed into a single commit, and then ./Tools/gittools/git-subsystems-split run on the resulting single commit.
The "apply patch" part of your fix, I dont like very much tho, and I think we should do everything we can to avoid doing that.

I think this commit is "really close".

arg7 commented:
Hi folks, I've rebased this branch on recent master, removed stuff already merged, cancelled patch of soc.h system library, squashed and split commits properly and ready for merge in master.

tpwrules reviewed (approved):
Force pushed to actually delete the old sdkconfig.

Looks good, I got it running and flying on my Stamp Fly. Had to tweak some stuff with I2C but I understand a future patch is coming, and it's for a not-in-tree driver anyway.

I'm not sure it flies quite as well as master but I think I was having some hardware issues too.
